src/pages/TokenDetails/Widget.tsx

```diff
@@ -0,0 +1,183 @@
+  useUserTransactionTTL,
+} from 'state/user/hooks'
+import { DARK_THEME, LIGHT_THEME } from 'theme/widget'
+
+export const WIDGET_WIDTH = 320
+
+const WIDGET_ROUTER_URL = 'https://api.uniswap.org/v1/'
+
+export interface WidgetProps {
+  defaultToken: Currency
+}
+
+export default function Widget({ defaultToken }: WidgetProps) {
+  const locale = useActiveLocale()
+  const darkMode = useIsDarkMode()
+  const theme = useMemo(() => (darkMode ? DARK_THEME : LIGHT_THEME), [darkMode])
+  const { provider } = useWeb3React()
+
+  const { value, tokenSelector } = useSyncWidgetValue(defaultToken)
+  const { settings } = useSyncWidgetSettings()
+  const { transactions } = useSyncWidgetTransactions()
+
+  return (
+    <>
+      <SwapWidget
+        hideConnectionUI
+        jsonRpcUrlMap={RPC_URLS}
+        routerUrl={WIDGET_ROUTER_URL}
+        width={WIDGET_WIDTH}
+        locale={locale}
+        theme={theme}
+        provider={provider}
+        {...value}
+        {...settings}
+        {...transactions}
+      />
+      {tokenSelector}
+    </>
+  )
+}
+
+// Integrates the Widget's settings, keeping the widget and app settings in sync.
+function useSyncWidgetSettings() {
+  const [userTtl, setUserTtl] = useUserTransactionTTL()
+  const [ttl, setTtl] = useState<number | undefined>(userTtl / 60)
+  const onTransactionDeadlineChange = useCallback(
+    (ttl: number | undefined) => {
+      setTtl(ttl)
+      const userTtl = ttl === undefined ? ttl : ttl * 60
+      setUserTtl(userTtl ?? DEFAULT_DEADLINE_FROM_NOW)
+    },
+    [setUserTtl]
+  )
+  const [userSlippage, setUserSlippage] = [useUserSlippageTolerance(), useSetUserSlippageTolerance()]
+  const [slippage, setSlippage] = useState<string | undefined>(
+    userSlippage === 'auto' ? undefined : userSlippage.toFixed(2)
+  )
+  const onSlippageChange = useCallback(
+    (slippage: Slippage) => {
+      setSlippage(slippage.max)
+      if (slippage.auto || !slippage.max) {
+        setUserSlippage('auto')
+      } else {
+        setUserSlippage(new Percent(Math.floor(Number(slippage.max) * 100), 10_000))
+      }
+    },
+    [setUserSlippage]
+  )
+  const onSettingsReset = useCallback(() => {
+    setTtl(undefined)
+    setUserTtl(DEFAULT_DEADLINE_FROM_NOW)
+    setSlippage(undefined)
+    setUserSlippage('auto')
+  }, [setUserSlippage, setUserTtl])
+  const settings: SwapSettingsController = useMemo(() => {
+    const auto = userSlippage === 'auto'
+    return { slippage: { auto, max: slippage }, transactionTtl: ttl }
+  }, [slippage, ttl, userSlippage])
+  const settingsHandlers: SwapEventHandlers = useMemo(
+    () => ({ onSettingsReset, onSlippageChange, onTransactionDeadlineChange }),
+    [onSettingsReset, onSlippageChange, onTransactionDeadlineChange]
+  )
+
+  return { settings: { settings, ...settingsHandlers } }
+}
+
+// Integrates the Widget's controlled value, using the app's CurrencySearchModal for token selection.
+function useSyncWidgetValue(defaultToken: Currency) {
```

Comment:
i would use `useSyncWidgetInputs`

Reply:
this function is doing a lot. would recommend splitting it into smaller helper hooks or something?